src/v/storage/segment_utils.cc

```diff
@@ -567,14 +576,14 @@ ss::future<ss::lw_shared_ptr<segment>> make_concatenated_segment(
       std::nullopt);
 }
 
+// assumes write_lock on `to` segment
```

Comment:
why not pass in the `lock::holder` resources, specially if it is a public call.

Reply:
well the caller may want to hold the locks for longer than the duration of this function. BUT i think this is an example of where we are seeing coroutines influence design patterns. with coroutines i can hold locks in my frame across an arbitrary series of async operations. Without coroutines i would need to pass them down through a nesting of continuations like you're alluding to. It does alter the way we compose things. Not sure what we want to do about this.

Reply:
The latter makes it easy to combine with existing code in the future without much refactoring. But yeah, that's how the rest of the calls in storage are afaik. your call.

Reply:
Though in this case, I think it makes easy to do the wrong thing by not asking for lockholder

Reply:
So is your proposal for a design pattern for co-routines to be something like

```
using locks_t = std::vector<ss::rwlock::holder>;
ss::future<locks_t> method(locks_t&& locks) {
   ...
   return std::move(locks);
}
```

this pattern allows composition at the expense of not being useful if something is used exclusively in a coroutine context.

Reply:
Seems right to me.

Reply:
So in a purely coroutine context it would look like this:

```
locks_t locks = ...;
locks = co_await method(std::move(locks));
locks = co_await method(std::move(locks));
...
```

weird if there aren't non-coroutine users?

Reply:
Sure, my only comment is that we should design interfaces that make it easy to do the right thing and hard to do the wrong thing. 

It's probably not the case that it will look like that in practice tho, but yes in the general case. 

Though you could pass a lw_shared_ptr<locks> around in this particular case so it will look like

`co_await foo(ptr); co_await bar(ptr); co_await baz(ptr)`